File: aoc-clj-2021/src/aoc_clj_2021/day11.clj
(ns aoc-clj-2021.day11
  (:require [clojure.string :as str]
            [clojure.set :as set]
            [clojure.java.io :as io]))

(def data (->> (io/resource "day11.txt") slurp str/trim-newline str/split-lines))
(def octopi (->> data (mapv vec) (mapv (fn [v] (mapv (comp parse-long str ) v)))))

(defn inc-by-one [[octopi flashed?]]
  [(mapv #(mapv inc %) octopi)
   flashed?])

(defn neighbors [loc]
  (for [i [-1 0 1] j [-1 0 1] :when (not= [i j] [0 0])]
    (mapv + loc [i j])))

(defn flash-step [[octopi flashed?]]
  (let [will-flash (set (for [[r row] (map-indexed vector octopi)
                              [c col] (map-indexed vector row)
                              :when (and (> col 9) (not (flashed? [r c])))]
                          [r c]))
        neighbors (->> will-flash (mapcat neighbors)  (remove #(nil? (get-in octopi %))))
        oct (reduce #(update-in %1 %2 inc) octopi neighbors)]
    [oct (set/union will-flash flashed?)]))

(defn wont-flash [octopi]
  (->> (keep-indexed (fn [r row]
                       (keep-indexed (fn [c col] (when (<= col 9) [r c]))
                                     row))
                     octopi)
       (apply concat)
       (set)))

(defn flash-step-done? [[octopi flashed?]]
  (= 100 (count (set/union (wont-flash octopi) flashed?))))

(defn reset-flashed [[octopi flashed? total-flashed]]
  [(reduce #(assoc-in %1 %2 0) octopi flashed?)
   #{}
   (+ total-flashed (count flashed?))])

(defn full-step [[octopi flashed? total-flashed :as state]]
  (let [step (->> [octopi flashed?]
                  (inc-by-one)
                  (iterate flash-step)
                  (some #(when (flash-step-done? %) %)))]
    (reset-flashed (conj step total-flashed))))

;; part 1
(let [[_ _ total-flashes] (nth (iterate full-step [octopi #{} 0]) 100)]
  total-flashes)

;; part 2
(some (fn [[i [octopi _ _]]]
        (when (->> octopi flatten (every? zero?)) i))
      (map vector
           (iterate inc 0)
           (iterate full-step [octopi #{} 0])))

File: aoc-py-2021/day11.py
data = open('day11.txt').read().splitlines()
nums = [list(map(int,line)) for line in data]

def in_bounds(loc): 
    return 0 <= loc[0] < len(nums) and 0 <= loc[1] < len(nums[0])

def neighbor_locs(loc):
    moves = [(r,c) for r in [0,1,-1] for c in [0,1,-1] if not (r,c) == (0,0)]
    return tuple(tuple(map(lambda x,y:x+y,m,loc)) for m in moves)

def incr_by_one(octopi):
    return  [list(map(lambda x: x+1,row)) for row in octopi]

def none_will_flash(octopi,flashed_locs):
    return all((r,c) in flashed_locs or col <= 9
              for r,row in enumerate(octopi) 
              for c,col in enumerate(row))

def reset_flashed(octopi,flashed_locs):
    return [[col if (r,c) not in flashed_locs else 0 
             for c,col in enumerate(row)]
             for r,row in enumerate(octopi)]
    
def flash(octopi,flashed_locs=set(),nflashed=0):
    if none_will_flash(octopi,flashed_locs):
        return [reset_flashed(octopi,flashed_locs),nflashed]
    flashed = set()
    for i,row in enumerate(octopi):
        for j,col in enumerate(row):
            if col > 9 and not (i,j) in flashed_locs:
                flashed.add((i,j))
                nlocs = tuple(n for n in neighbor_locs([i,j]) if in_bounds(n))
                
                for r,c in nlocs:
                    octopi[r][c] += 1                    
    return flash(octopi,flashed_locs.union(flashed),nflashed+len(flashed))

def step(octopi):
    return flash(incr_by_one(octopi))

# part 1
octopi,flashed = nums,0
for i in range(100):
    octopi,flashes = step(octopi)
    flashed += flashes 
print(flashed)

# part 2
i = 0
octopi = nums
while 1:
    octopi,flashes = step(octopi)
    i += 1
    if flashes == 100:
        break
print(i)